File: LeetCode/3.9/maximumBeauty.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
#define INDEX_PRICE 0
#define INDEX_BEAUTY 1

// int* maximumBeauty(int** items, int itemsSize, int* itemsColSize, int* queries, int queriesSize, int* returnSize) {
//     *returnSize=queriesSize;

//     int *ret=(int *)malloc(sizeof(int)*queriesSize);
//     for(int i=0;i<queriesSize;i++){
//         int max=0;
//         for(int j=0;j<itemsSize;j++){
//             if(queries[i]>=items[j][INDEX_PRICE]){
//                 max=max>items[j][INDEX_BEAUTY]?max:items[j][INDEX_BEAUTY];
//             }
//         }
//         ret[i]=max;
//     }
//     return ret;
// }

// 这么写包超时的

int compare(const void *a, const void *b)
{
    return (*(int **)a)[INDEX_PRICE] - (*(int **)b)[INDEX_PRICE];
}

int *maximumBeauty(int **items, int itemsSize, int *itemsColSize, int *queries, int queriesSize, int *returnSize)
{
    *returnSize = queriesSize;
    int *ret = (int *)malloc(sizeof(int) * queriesSize);
    int *maxBeauty = (int *)malloc(sizeof(int) * itemsSize);

    qsort(items, itemsSize, sizeof(int *), compare);

    maxBeauty[0] = items[0][INDEX_BEAUTY];
    for (int i = 1; i < itemsSize; i++)
    {
        maxBeauty[i] = maxBeauty[i - 1] > items[i][INDEX_BEAUTY] ? maxBeauty[i - 1] : items[i][INDEX_BEAUTY];
    }

    for (int i = 0; i < queriesSize; i++)
    {
        int left = 0;
        int right = itemsSize - 1;
        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            if (queries[i] >= items[mid][INDEX_PRICE])
            {
                left = mid + 1;
            }
            else
            {
                right = mid - 1;
            }
        }
        ret[i] = right >= 0 ? maxBeauty[right] : 0;
    }

    free(maxBeauty);
    return ret;
}

File: LeetCode/25年8月/8.20/maximumBeauty.java

import java.util.Arrays;

class Solution {

    public int[] maximumBeauty(int[][] items, int[] queries) {
        int len = queries.length;
        int[] ans = new int[len];

        Arrays.sort(items, (a, b) -> Integer.compare(a[0], b[0]));//按price升序

        // 预处理：计算前缀最大美丽值
        for (int i = 1; i < items.length; i++) {
            items[i][1] = Math.max(items[i][1], items[i - 1][1]);
        }
        for (int i = 0; i < len; i++) {
            int index = BinarySearch(items, queries[i] + 1);
            if (index == 0) {
                ans[i] = 0;
            } else {
                ans[i] = items[index - 1][1]; // 直接获取前缀最大值
            }
        }

        // for (int i = 0; i < len; i++) {
        //     int index = BinarySearch(items, queries[i] + 1);
        //     if (index == 0) {
        //         ans[i] = 0;
        //     } else {
        //         for (int j = 0; j < index; j++) {
        //             ans[i] = ans[i] > items[j][1] ? ans[i] : items[j][1];
        //         }
        //     }
        // }
        return ans;
    }

    private int BinarySearch(int[][] items, int queries) {
        int left = 0;
        int right = items.length;
        int mid;
        while (left < right) {
            mid = left + (right - left) / 2;
            if (items[mid][0] < queries) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}
